README
README
======

Red Feline Thermometer is a USB termperature sensor. It is designed to be
used attached to a computer for logging applications. It can be fitted into
small enclosures such as a 35mm film canister which can be easily weather
proofed in order to measure outside temperatures. The associated host
program can be used to perform semi-automatic single ended calibration or,
for greater accuracy across a wide temperature range, manual two point
calibration. The host program can also read the data from the device
presenting on standard output allowing additional tools to be easily
written to log data in any manner desired. 

It could be used:

- to maintain long term records the outside (or inside) temperature,
- to graph the heating/cooling cycle of domestic heating systems,
- to compare the heating/cooling cycles of different rooms in the houses (in order to tune the radiator valves),
- to bore people at parties.


FILES IN THE DISTRIBUTION
-------------------------

Readme.txt ........ The file you are currently reading.
circuit/ .......... Diagrams of the circuit.
driver/ ........... Source code for the thermometer manager (thermoman)
firmware/ ......... Source code of the controller firmware.
firmware/usbdrv/ .. USB driver -- See Readme.txt in that directory for info
COPYING ........... Public license (GPL2) for all contents of this project.
Changelog ......... Logfile documenting changes in firm- and hardware.


COMPILING AND INSTALLING
------------------------

This project can be built on Unix (GNU/Linux, FreeBSD or Mac OS X) or Windows
although no testing has been done on any platform except GNU/Linux.

Building on Windows:

You need WinAVR to compile the firmware. A package can be downloaded from:

    WinAVR: http://winavr.sourceforge.net/

To build the firmware with WinAVR, change into the "firmware" directory,
check whether you need to edit the "Makefile" (e.g. change the ISP upload
tool) and type "make" to compile the source code. Before you upload the code
to the device with "make flash", you should set the fuses with "make fuse".

Building on Unix (GNU/Linux, FreeBSD and Mac):

You need the GNU toolchain and avr-libc to compile the firmware. See

    http://www.nongnu.org/avr-libc/user-manual/install_tools.html

for instructions on how to install avr-gcc and avr-libc.

To build the firmware, change to the "firmware" directory, edit "Makefile"
to use the programmer of your choice and type "make" to compile the source
code. Before you upload the code to the device with "make flash", you
should set the fuses with "make fuse".

To build the driver, change to the "driver" directory and type "make". The
driver program, thermoman will be built and can be run from the "driver"
directory.


USING THERMOMAN
---------------

Thermoman uses GNU long options and a summary of usage can be obtained
using "thermoman --help":

--- cut here ---
Usage: ./thermoman [OPTION]...
Driver for the RFThermometer. Shows the current temperature by
default but can also be used to calibrate the device using both
single and double ended methods

Mandatory arguments to long options are mandatory for short options too.
  -c, --calibrate=SENSOR:MIN:MAX:MINADC:MAXADC
                             calibrate a specific sensor using the supplied
                             range (e.g. --calibrate=2v:20:52:440:508 ).
                             Sensor names are: int, 1v, 2v and 3v
      --factory-calibration  adopt factory calibration
  -h, --help                 display this help and exit
  -r, --reset                reset the min/max values
  -s, --show=SENSOR          show results measured with SENSOR (default: 3v)
  -H, --show-hex             show the raw hex packet exchanged with hardware
  -R, --show-raw             show the raw decimal values read from the h/ware
      --single-ended=TEMP    perform fast single ended calibration based on
                             the current temperature (leave time to warm up)
      --unsafe               enable operations that write to devices EEPROM
--- cut here ---

The quick start guide is simple to run "thermoman" with no arguments. If the
sensor has been calibrated the curent temperature (together with the highest
and lowest values observed since power on) will be displayed.

If the sensor has not yet been calibrated (for example because you've only
just built it then a quick and dirty single ended calibration can be
performed from another temperature sensor by calling:
  thermoman --single-ended=<value_in_celcius> --unsafe
  
Note: Leave the device powered up for at least twenty minutes before
      calibrating. Many builds exhibit very significant drift during
      warm up.


HARDWARE
--------

RFThermometer is a member of a family of simple platforms contructed from
Atmel's ATTiny85 microcontrollers. All family members share the same USB
pins and these pins have been carefully selected to avoid using pins
that are needed for in-circuit programming (ICSP). This allows them to be
reprogrammed without disconnecting them from the computer that supplies
them power (if they have an ICSP connector)

The RFThermometer differs from the standard platform by removing the ICSP
connector and using a narrower circuit board allowing it to be squeezed
into smaller enclosures.

USB communications using software bit-banging (thanks to Objective
Development's V-USB driver). Strictly speaking this circuit does not
meeting the USB electrical specification but is close enough to work with
almost all USB hosts. It also uses the USB host clock to calibrate the
ATTiny's internal oscillator eliminating the need for an external crystal.

The result is a very cheap circuit to build. Have fun.


MORE INFORMATION
----------------

RFThermometer was written by Daniel Thompson using driver software and
example code from Objective Development. For other projects by the same
author see:

    http://www.redfelineninja.org.uk/

For more information about Objective Development's firmware-only USB driver
for Atmel's AVR microcontrollers please visit the URL

    http://www.obdev.at/products/vusb/

A technical documentation of the driver's interface can be found in the
file "firmware/usbdrv/usbdrv.h".


ChangeLog
ChangeLog
=========

0.6:
- Rename hidtool to thermoman (thermometer manager)
- Update the README.

0.5:
- Initial release.



firmware/main.c
/*
 * main.c
 *
 * Main functions for RFThermometer, a firmware for a USB thermometer based
 * on vusb.
 *
 * Copyright (C) 2011 Daniel Thompson <daniel@redfelineninja.org.uk>
 * Copyright (C) 2008 by OBJECTIVE DEVELOPMENT Software GmbH
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */


#include <avr/io.h>
#include <avr/wdt.h>
#include <avr/eeprom.h>
#include <avr/interrupt.h>
#include <avr/pgmspace.h>
#include <util/delay.h>

#include <stdlib.h>
#include <string.h>

#include "usbdrv.h"
#include "oddebug.h"

#include "osccal.h"

/* ------------------------------------------------------------------------- */

#define lengthof(x) (sizeof(x) / sizeof(x[0]))

#define BUTTON_PORT PORTB       /* PORTx - register for button output */
#define BUTTON_PIN PINB         /* PINx - register for button input */
#define BUTTON_BIT PB0          /* bit for button input/output */

#define LED_BIT PB1
#define LED_ON()   (PORTB &= ~_BV(LED_BIT))
#define LED_OFF()  (PORTB |= _BV(LED_BIT))
#define LED_TOGGLE() (PORTB ^= _BV(LED_BIT))
#define LED_INIT() (LED_OFF(), DDRB |= _BV(LED_BIT))

#define UTIL_BIN4(x)        (uchar)((0##x & 01000)/64 + (0##x & 0100)/16 + (0##x & 010)/4 + (0##x & 1))
#define UTIL_BIN8(hi, lo)   (uchar)(UTIL_BIN4(hi) * 16 + UTIL_BIN4(lo))

/* ------------------------------------------------------------------------- */

static const uchar conversionTable[] = {
	UTIL_BIN8(1000, 1111), /* Vref=1.1V, ADC4 (temp sensor) */
	UTIL_BIN8(1000, 0001), /* Vref=1.1V, ADC1 (PB2) */
	UTIL_BIN8(1001, 0001), /* Vref=2.56V, ADC1 (PB2) */
	UTIL_BIN8(0000, 0001), /* Vref=3.3V, ADC1 (PB2) */
};

static volatile struct {
	unsigned int current[lengthof(conversionTable)];
	unsigned int min[lengthof(conversionTable)];
	unsigned int max[lengthof(conversionTable)];
} conversion;

/* ------------------------------------------------------------------------- */
/* ----------------------------- USB interface ----------------------------- */
/* ------------------------------------------------------------------------- */

PROGMEM char usbHidReportDescriptor[22] = {    /* USB report descriptor */
    0x06, 0x00, 0xff,              // USAGE_PAGE (Generic Desktop)
    0x09, 0x01,                    // USAGE (Vendor Usage 1)
    0xa1, 0x01,                    // COLLECTION (Application)
    0x15, 0x00,                    //   LOGICAL_MINIMUM (0)
    0x26, 0xff, 0x00,              //   LOGICAL_MAXIMUM (255)
    0x75, 0x08,                    //   REPORT_SIZE (8)
    0x95, 0x80,                    //   REPORT_COUNT (128)
    0x09, 0x00,                    //   USAGE (Undefined)
    0xb2, 0x02, 0x01,              //   FEATURE (Data,Var,Abs,Buf)
    0xc0                           // END_COLLECTION
};
/* Since we define only one feature report, we don't use report-IDs (which
 * would be the first byte of the report). The entire report consists of 128
 * opaque data bytes.
 */

/* The following variables store the status of the current data transfer */
static uchar    currentAddress;
static uchar    bytesRemaining;

/* ------------------------------------------------------------------------- */

static uchar usbGetSensorReport(uchar offset) {
	if (offset == 1)
		return OSCCAL;

	if (offset >= 2 && offset < 26)
		return ((uchar *) (&conversion))[offset - 2];

	return eeprom_read_byte((uchar *)0 + offset);
}


/* usbFunctionRead() is called when the host requests a chunk of data from
 * the device. For more information see the documentation in usbdrv/usbdrv.h.
 */
uchar   usbFunctionRead(uchar *data, uchar len)
{
    uchar i;

    if(len > bytesRemaining)
        len = bytesRemaining;

    for (i=0; i<len; i++, currentAddress++)
	    data[i] = usbGetSensorReport(currentAddress);
    bytesRemaining -= len;
	 
    return len;
}

/* usbFunctionWrite() is called when the host sends a chunk of data to the
 * device. For more information see the documentation in usbdrv/usbdrv.h.
 */
uchar   usbFunctionWrite(uchar *data, uchar len)
{
    if(bytesRemaining == 0)
        return 1; /* end of transfer */

    if(len > bytesRemaining)
        len = bytesRemaining;

    /* handle the nop and RAM based addresses */
    while (len != 0 && currentAddress < 26) {
        /* ignore 0 and 1 (write not allowed to those) */
	
	if (currentAddress >= 2)	    
	    ((uchar *) (&conversion))[currentAddress - 2] = *data;

	data++;
	len--;
	currentAddress++;
	bytesRemaining--;    
    }

    /* program the EEPROM based addresses. the driver always sends 128 bytes
     * (i.e. tries to alters EEPROM addresses) even when no changes have been
     * made so we must use eeprom_update_block() to prevent excessive EEPROM
     * wear.
     */
    if (len != 0) {
	    eeprom_update_block(data, (uchar *)0 + currentAddress, len);

	    currentAddress += len;
	    bytesRemaining -= len;
    }

    /* return 1 if this was the last chunk */
    return bytesRemaining == 0;
}

/* ------------------------------------------------------------------------- */

usbMsgLen_t usbFunctionSetup(uchar data[8])
{
usbRequest_t    *rq = (void *)data;

    if((rq->bmRequestType & USBRQ_TYPE_MASK) == USBRQ_TYPE_CLASS){    /* HID class request */
        if(rq->bRequest == USBRQ_HID_GET_REPORT){  /* wValue: ReportType (highbyte), ReportID (lowbyte) */
            /* since we have only one report type, we can ignore the report-ID */
            bytesRemaining = 128;
            currentAddress = 0;
            return USB_NO_MSG;  /* use usbFunctionRead() to obtain data */
        }else if(rq->bRequest == USBRQ_HID_SET_REPORT){
            /* since we have only one report type, we can ignore the report-ID */
            bytesRemaining = 128;
            currentAddress = 0;
            return USB_NO_MSG;  /* use usbFunctionWrite() to receive data from host */
        }
    }else{
        /* ignore vendor type requests, we don't use any */
    }
    return 0;
}

void hadUsbReset(void)
{
    const uchar threshold = 2;

    uchar oldcal, newcal;

    cli();
    calibrateOscillator();
    sei();

    oldcal = eeprom_read_byte(0);
    newcal = OSCCAL;

    /* store the calibrated value in EEPROM if it has changed by more
     * than the threshold
     */    
    if (((oldcal > newcal) && ((oldcal - newcal) > threshold)) ||
	((oldcal < newcal) && ((newcal - oldcal) > threshold)))
        eeprom_write_byte(0, OSCCAL);
}

/* ------------------------------------------------------------------------- */

#define ADC_START_CONVERSION() (ADCSRA |= _BV(ADSC))
#define ADC_IS_BUSY()  (ADCSRA & _BV(ADSC))

static void adcInit(void)
{
	memset((void *) conversion.min, 0xff, sizeof(conversion.min));

	/* enable ADC, not free running, interrupt disable, rate = 1/128 */	
	ADCSRA = UTIL_BIN8(1000, 0111);

	/* kick off the first conversion */
	ADMUX = conversionTable[0];
	ADC_START_CONVERSION();
}

static void adcPoll(void)
{
	static uchar state = 0;
	static uchar keepResult = 0;

	if (!ADC_IS_BUSY()) {
		if (!keepResult) {
			/* discard this result and start again */
			ADC_START_CONVERSION();
			keepResult = 1;
		} else {
			/* conversion is complete */
			if (!bytesRemaining) {
				conversion.current[state] = ADC;
				
				if (conversion.min[state] > ADC)
					conversion.min[state] = ADC;

				if (conversion.max[state] < ADC)
					conversion.max[state] = ADC;
			}
			
			/* whether we did the write or not if the latched value
			 * matches we're done
			 */
			if (conversion.current[state] == ADC) {
				/* switch to next state */
				state++;
				if (state >= lengthof(conversionTable))
					state = 0;
				
				/* reprogram the ADC (making sure we discard
				 * the next value)
				 */
				ADMUX = conversionTable[state];
				ADC_START_CONVERSION();
				keepResult = 0;
			}
		}
	}
}

/* ------------------------------------------------------------------------- */
/* --------------------------------- main ---------------------------------- */
/* ------------------------------------------------------------------------- */

int main(void)
{
uchar   i;
uchar   calibrationValue;

    calibrationValue = eeprom_read_byte(0); /* calibration value from last time */
    if(calibrationValue != 0xff){
        OSCCAL = calibrationValue;
    }
    
    usbInit();
    adcInit();

    usbDeviceDisconnect();  /* enforce re-enumeration, do this while interrupts are disabled! */
    i = 0;
    while(--i){             /* fake USB disconnect for > 250 ms */
        wdt_reset();
        _delay_ms(1);
    }
    usbDeviceConnect();

    wdt_enable(WDTO_1S);

    LED_INIT();
	/* turn on internal pull-up resistor for the switch */
    BUTTON_PORT |= _BV(BUTTON_BIT);

    sei();

    for(;;){    /* main event loop */
        wdt_reset();
        usbPoll();
	adcPoll();
    }

    return 0;
}

/* ------------------------------------------------------------------------- */


driver/thermoman.c
/*
 * thermoman.c
 *
 * Main functions for thermoman, a driver for a USB thermometer
 *
 * Copyright (C) 2011 Daniel Thompson <daniel@redfelineninja.org.uk>
 * Copyright (C) 2008 OBJECTIVE DEVELOPMENT Software GmbH
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */

#define _GNU_SOURCE

#include <assert.h>
#include <ctype.h>
#include <getopt.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>
#include <stdint.h>
#include "hiddata.h"
#include "../firmware/usbconfig.h"  /* for device VID, PID, vendor name and product name */

/* ------------------------------------------------------------------------- */

typedef enum {
	ID_INTERNAL,
	ID_EXTERNAL_1_1,
	ID_EXTERNAL_2_56,
	ID_EXTERNAL_3_3,
	ID_MAX,
	ID_INVALID,
} rfsense_id_t;

typedef struct {
	uint16_t minadc;
	uint16_t minunit;
	uint16_t maxadc;
	uint16_t maxunit;
} rfsense_calibration_t;

typedef struct {
	uint8_t osccal_boot;
	uint8_t osccal_actual;

	uint16_t current[ID_MAX];
	uint16_t min[ID_MAX];
	uint16_t max[ID_MAX];

	rfsense_calibration_t calib[ID_MAX];
} rfsense_state_t;

typedef struct {
	double current[ID_MAX];
	double min[ID_MAX];
	double max[ID_MAX];
} rfsense_units_t;

/* ------------------------------------------------------------------------- */

rfsense_calibration_t rfsense_volts[ID_MAX] = {
	{ 0, 0, 0x3ff, 110 },
	{ 0, 0, 0x3ff, 110 },
	{ 0, 0, 0x3ff, 256 },
	{ 0, 0, 0x3ff, 330 }
};

rfsense_calibration_t rfsense_temp[ID_MAX] = {
	{ 300, 2500, 370, 9500 }, /* ATTinyx5 datasheet */
	{ 372,    0, 934, 3100 }, /* MCP9701(A) */
	{ 160,    0, 401, 3100 }, /* MCP9701(A) */
	{ 124,    0, 311, 3100 }  /* MCP9701(A) */
};

double rfsense_get_unit(uint16_t adc, const rfsense_calibration_t *calib)
{
	double x = adc;

	double x1 = calib->minadc;
	double y1 = calib->minunit / 100;
	double x2 = calib->maxadc;
	double y2 = calib->maxunit / 100;
	
	double m = -(y2 - y1) / (x1 - x2);
	double c = -((m*x1) - y1);

	return m*x + c;
}

void rfsense_decode(const uint8_t *report, rfsense_state_t *state)
{
#define R8(offset) (report[offset])
#define R16(offset) ((R8(offset+1) << 8) | R8(offset))

	int i;

	state->osccal_boot = R8(0);
	state->osccal_actual = R8(1);

	for (i=0; i<4; i++) {
		state->current[i] = R16(2 + 2*i);

		state->min[i] = R16(10 + 2*i);
		state->max[i] = R16(18 + 2*i);

		state->calib[i].minadc = R16(26 + 8*i);
		state->calib[i].minunit = R16(28 + 8*i);
		state->calib[i].maxadc = R16(30 + 8*i);
		state->calib[i].maxunit = R16(32 + 8*i);
	}
#undef R8
#undef R16
}

void rfsense_state_to_units(const rfsense_state_t *state,
			    const rfsense_calibration_t *calib,
			    rfsense_units_t *units)
{
	int i;

	if (NULL == calib)
		calib = state->calib;

	for (i=0; i<ID_MAX; i++) {
		units->current[i] = rfsense_get_unit(state->current[i], &calib[i]);
		units->min[i] = rfsense_get_unit(state->min[i], &calib[i]);
		units->max[i] = rfsense_get_unit(state->max[i], &calib[i]);
	}		
}

void rfsense_encode(const rfsense_state_t *state, uint8_t *report)
{
#define W8(offset, val) (report[offset] = val)
#define W16(offset, val) (W8(offset+1, val >> 8), W8(offset, val & 0xff))
	int i;

	W8(0, state->osccal_boot);
	W8(1, state->osccal_actual);

	for (i=0; i<4; i++) {
		W16(2 + 2*i, state->current[i]);

		W16(10 + 2*i, state->min[i]);
		W16(18 + 2*i, state->max[i]);
		
		W16(26 + 8*i, state->calib[i].minadc);
		W16(28 + 8*i, state->calib[i].minunit);
		W16(30 + 8*i, state->calib[i].maxadc);
		W16(32 + 8*i, state->calib[i].maxunit);
	}	
#undef W8
#undef W16
}

/**
 * Perform single ended calibration using the current temperature as the ADC reference.
 */
void rfsense_set_temp(rfsense_state_t *state, int temp)
{
	temp *= 100;

	state->calib[ID_INTERNAL].minunit = temp;
	state->calib[ID_INTERNAL].minadc = state->current[ID_INTERNAL];
	state->calib[ID_INTERNAL].maxunit = temp + 2000;
	state->calib[ID_INTERNAL].maxadc = state->current[ID_INTERNAL] + 20;

	// this one can overflow easily so we go down rather than up
	state->calib[ID_EXTERNAL_1_1].minunit = temp - 800;
	state->calib[ID_EXTERNAL_1_1].minadc = state->current[ID_EXTERNAL_1_1] - 145;
	state->calib[ID_EXTERNAL_1_1].maxunit = temp;
	state->calib[ID_EXTERNAL_1_1].maxadc = state->current[ID_EXTERNAL_1_1];

	state->calib[ID_EXTERNAL_2_56].minunit = temp;
	state->calib[ID_EXTERNAL_2_56].minadc = state->current[ID_EXTERNAL_2_56];
	state->calib[ID_EXTERNAL_2_56].maxunit = temp + 2200;
	state->calib[ID_EXTERNAL_2_56].maxadc = state->current[ID_EXTERNAL_2_56] + 171;

	state->calib[ID_EXTERNAL_3_3].minunit = temp;
	state->calib[ID_EXTERNAL_3_3].minadc = state->current[ID_EXTERNAL_3_3];
	state->calib[ID_EXTERNAL_3_3].maxunit = temp + 3100;
	state->calib[ID_EXTERNAL_3_3].maxadc = state->current[ID_EXTERNAL_3_3] + 187;
}

void _rfsense_print_raw(uint16_t *adc, double *unit, FILE *fp)
{
	fprintf(fp, "%5d (%5.2f)  ", adc[ID_INTERNAL], unit[ID_INTERNAL]);
	fprintf(fp, "%5d (%5.2f)  ", adc[ID_EXTERNAL_1_1], unit[ID_EXTERNAL_1_1]);
	fprintf(fp, "%5d (%5.2f)  ", adc[ID_EXTERNAL_2_56], unit[ID_EXTERNAL_2_56]);
	fprintf(fp, "%5d (%5.2f)\n", adc[ID_EXTERNAL_3_3], unit[ID_EXTERNAL_3_3]);
}

void rfsense_print_raw(rfsense_state_t *state, FILE *fp)
{
	rfsense_units_t units;
	rfsense_state_to_units(state, NULL, &units);


	fprintf(fp, "OSCCAL:     Actual 0x%02x    Stored 0x%02x\n",
		state->osccal_actual, state->osccal_boot);

	fprintf(fp, "Current:  ");
	_rfsense_print_raw(state->current, units.current, fp);

        fprintf(fp, "Minimum:  ");
	_rfsense_print_raw(state->min, units.min, fp);

	fprintf(fp, "Maximum:  ");
	_rfsense_print_raw(state->max, units.max, fp);
}

/**
 * Display the results from a sensor.
 *
 * Note the, unlike the raw results, we only show one decimal place.
 * This is to avoid "tricking" the user about the precision offered
 * by the hardware.
 */
void rfsense_print(rfsense_state_t *state, rfsense_id_t id, FILE *fp)
{
	assert(id < ID_MAX);

	rfsense_units_t units;
	rfsense_state_to_units(state, NULL, &units);

	fprintf(fp, "%4.1f        Min %4.1f    Max %4.1f\n",
			units.current[id], units.min[id], units.max[id]);
}

/* ------------------------------------------------------------------------- */

static char *usbErrorMessage(int errCode)
{
static char buffer[80];

    switch(errCode){
        case USBOPEN_ERR_ACCESS:      return "Access to device denied";
        case USBOPEN_ERR_NOTFOUND:    return "The specified device was not found";
        case USBOPEN_ERR_IO:          return "Communication error with device";
        default:
            sprintf(buffer, "Unknown USB error %d", errCode);
            return buffer;
    }
    return NULL;    /* not reached */
}

static usbDevice_t  *openDevice(void)
{
usbDevice_t     *dev = NULL;
unsigned char   rawVid[2] = {USB_CFG_VENDOR_ID}, rawPid[2] = {USB_CFG_DEVICE_ID};
char            vendorName[] = {USB_CFG_VENDOR_NAME, 0}, productName[] = {USB_CFG_DEVICE_NAME, 0};
int             vid = rawVid[0] + 256 * rawVid[1];
int             pid = rawPid[0] + 256 * rawPid[1];
int             err;

    if((err = usbhidOpenDevice(&dev, vid, vendorName, pid, productName, 0)) != 0){
        fprintf(stderr, "error finding %s: %s\n", productName, usbErrorMessage(err));
        return NULL;
    }
    return dev;
}

/* ------------------------------------------------------------------------- */

static void hexdump(uint8_t *buffer, int len)
{
int     i;
FILE    *fp = stdout;

    for(i = 0; i < len; i++){
        if(i != 0){
            if(i % 16 == 0){
                fprintf(fp, "\n");
            }else{
                fprintf(fp, " ");
            }
        }
        fprintf(fp, "0x%02x", buffer[i] & 0xff);
    }
    if(i != 0)
        fprintf(fp, "\n");
}

#if 0
static int  hexread(uint8_t *buffer, char *string, int buflen)
{
char    *s;
int     pos = 0;

    while((s = strtok(string, ", ")) != NULL && pos < buflen){
        string = NULL;
        buffer[pos++] = (char)strtol(s, NULL, 0);
    }
    return pos;
}
#endif

/* ------------------------------------------------------------------------- */



static void issue_help(char *myName)
{
	printf(
"Usage: %s [OPTION]...\n"
"Driver for the RFThermometer. Shows the current temperature by\n"
"default but can also be used to calibrate the device using both\n"
"single and double ended methods\n"
"\n"
"Mandatory arguments to long options are mandatory for short options too.\n"
"  -c, --calibrate=SENSOR:MIN:MAX:MINADC:MAXADC\n"
"                             calibrate a specific sensor using the supplied\n"
"                             range (e.g. --calibrate=2v:20:52:440:508 ).\n"
"                             Sensor names are: int, 1v, 2v and 3v\n"
"      --factory-calibration  adopt factory calibration\n"
"  -h, --help                 display this help and exit\n"
"  -r, --reset                reset the min/max values\n"
"  -s, --show=SENSOR          show results measured with SENSOR (default: 3v)\n"
"  -H, --show-hex             show the raw hex packet exchanged with hardware\n"
"  -R, --show-raw             show the raw decimal values read from the h/ware\n"
"      --single-ended=TEMP    perform fast single ended calibration based on\n"
"                             the current temperature (leave time to warm up)\n"
"      --unsafe               enable operations that write to devices EEPROM\n"
"\n",
	       myName);
}

static rfsense_id_t unpack_id(const char *sid)
{
	if (0 == strcmp(sid, "int")) {
		return ID_INTERNAL;
	} else if (0 == strcmp(sid, "1v")) {
		return ID_EXTERNAL_1_1;
	} else if (0 == strcmp(sid, "2v")) {
		return ID_EXTERNAL_2_56;
	} else if (0 == strcmp(sid, "3v")) {
		return ID_EXTERNAL_3_3;
	}

	return ID_INVALID;
}

static int unpack_calibration(const char *s, rfsense_id_t *idp, rfsense_calibration_t *calibration)
{
	char buf[80]; // this is very generous
	char delim[] = ":, ";
	int len;
	char *sid, *smin, *smax, *sminadc, *smaxadc, *snull;
	rfsense_id_t id;

	len = strlen(s) + 1;
	if (len > sizeof(buf))
		return -1;

	memcpy(buf, s, len);

	sid = strtok(buf, delim);
	smin = strtok(NULL, delim);
	smax = strtok(NULL, delim);
	sminadc = strtok(NULL, delim);
	smaxadc = strtok(NULL, delim);
	snull = strtok(NULL, delim);

	if (smaxadc == NULL || snull != NULL)
		return -1;

	if (!isdigit(smin[0]) || !isdigit(smax[0]) || !isdigit(sminadc[0]) || !isdigit(smaxadc[0]))
		return -1;

	id = unpack_id(sid);
	if (id == ID_INVALID)
		return -1;

	// success guaranteed after this point so wecan start modifying the
	// out parameters

	*idp = id;
	calibration->minunit = atof(smin) * 100.0;
	calibration->minadc = atoi(sminadc);
	calibration->maxunit = atof(smax) * 100.0;
	calibration->maxadc = atoi(smaxadc);

#if 0
	printf("%s -> %d   %s -> %d   %s -> %d   %s -> %d   %s -> %d\n",
		sid, *id,
		smin, calibration->minunit, sminadc, calibration->minadc,
		smax, calibration->maxunit, smaxadc, calibration->maxadc);
#endif
	return 0;
}

int main(int argc, char **argv)
{
	enum {
		OPT_FC = 256,
		OPT_SINGLE,
		OPT_UNSAFE,
	};

	static struct option long_options[] = {
		{ "calibrate", 1, 0, 'c' },
		{ "factory-calibration", 0, 0, OPT_FC },
		{ "help", 0, 0, 'h' },
		{ "reset", 0, 0, 'r' },
		{ "single-ended", 1, 0, OPT_SINGLE },
		{ "show", 1, 0, 's' },
		{ "show-hex", 0, 0, 'H' },
		{ "show-raw", 0, 0, 'R' },
		{ "unsafe", 0, 0, OPT_UNSAFE },
	};

	int opt;

	/* arguments */
	rfsense_id_t calibrate = ID_INVALID;
	bool factory_calibration = false;
	bool reset = false;
	rfsense_id_t show = ID_INVALID;
	bool show_raw = false;
	bool show_hex = false;
	bool single_ended = false;
	bool unsafe = false;

	int  calibration_point = 20;
	rfsense_calibration_t calibration;
	
	/* derived arguments */
	bool bad_args = false;
	bool do_write = false;

	usbDevice_t *dev;
	uint8_t     buffer[129];    /* room for dummy report ID */
	int len = sizeof(buffer);
	int         err;
	rfsense_state_t state;

	while ((opt = getopt_long(argc, argv, "chrs:HR", long_options, NULL)) != -1) {
		switch (opt) {
		case 'c': // --calibrate
			if (0 != unpack_calibration(optarg, &calibrate, &calibration)) {
				fprintf(stderr, "Bad calibration set: '%s'\n", optarg);
				bad_args = true;
			}
			break;

		case OPT_FC: // --factory-calibration
			factory_calibration = true;
			break;

		case 'h': // --help
			issue_help(argv[0]);
			exit(0);
			break;

		case 'r': // --reset
			reset = true;
			break;

		case 's': // --show=SENSOR
			show = unpack_id(optarg);
			if (ID_INVALID == show) {
				fprintf(stderr, "Bad sensor name: '%s'\n", optarg);
				bad_args = true;
			}
			break;

		case 'H': // --show-hex
			show_hex = true;
			break;

		case 'R': // --show-raw
			show_raw = true;
			break;

		case OPT_SINGLE: // --single-ended
			single_ended = true;
			if (isdigit(optarg[0])) {
				calibration_point = atoi(optarg);
			} else {
				fprintf(stderr, "Bad calibration point: '%s'\n", optarg);
				bad_args = true;
			}
			break;
	
		case OPT_UNSAFE: // --unsafe
			unsafe = true;
			break;

		default:
			fprintf(stderr, "Try '%s --help' for more information.\n",
				argv[0]);
			return 1;
		}
	}

	if (optind < argc && bad_args) {
		if (!bad_args)
			fprintf(stderr, "%s: Too many arguments\n", argv[0]);
		fprintf(stderr, "Try '%s --help' for more information.\n", argv[0]);
		return 2;
	}

	if((dev = openDevice()) == NULL)
		return 3;

	if((err = usbhidGetReport(dev, 0, (char *) buffer, &len)) != 0) {
		fprintf(stderr, "error reading data: %s\n", usbErrorMessage(err));
		return 4;
	}

	rfsense_decode(buffer+1, &state);

	if (reset) {
		memset(state.min, 0xff, sizeof(state.min));
		memset(state.max, 0, sizeof(state.max));
	       
		do_write = true;
	}

	if (factory_calibration) {
		if (!unsafe) {
			fprintf(stderr, "Factory calibration will destroy data, requires --unsafe\n");
		} else {
			memcpy(state.calib, rfsense_temp, sizeof(state.calib));
			do_write = true;
		}
	}

	if (single_ended) {
		if (!unsafe) {
			fprintf(stderr, "Single ended calibration will destroy data, requires --unsafe\n");
		} else {
			rfsense_set_temp(&state, calibration_point);
			do_write = true;
		}
	}

	if (calibrate != ID_INVALID) {
		if (!unsafe) {
			fprintf(stderr, "Calibration destroys existing values, requires --unsafe\n");
		} else {
			state.calib[calibrate] = calibration;
			do_write = true;
		}
	}

	if (do_write) {
		rfsense_encode(&state, buffer+1);

		if((err = usbhidSetReport(dev, buffer, sizeof(buffer))) != 0)
			fprintf(stderr, "error writing data: %s\n",
				usbErrorMessage(err));
	}

	// we've now finished interacting with the USB device
	usbhidCloseDevice(dev);

	// select a default display mode if not other activity was requested
	if (!do_write && !reset && !show_raw && !show_hex && show == ID_INVALID)
		show = ID_EXTERNAL_3_3;

	if (show_raw)
		rfsense_print_raw(&state, stdout);

	if (show_hex)
		hexdump(buffer+1, sizeof(buffer)-1);

	if (show != ID_INVALID)
		rfsense_print(&state, show, stdout);

	return 0;
}

/* ------------------------------------------------------------------------- */
